In [1]:
import pandas as pd
import numpy as np
from datatile.summary.df import DataFrameSummary 

pd.options.mode.use_inf_as_na = True
# pd.set_option('')

df2009 = pd.read_csv('data_incoming\customer_transactions_sample__Export_1_2009-10.csv')
df2010 = pd.read_csv('data_incoming\customer_transactions_sample__Export_2_2010-11.csv')


c:\Users\sraja\Miniconda3\lib\site-packages\numpy\_distributor_init.py:30: UserWarning: loaded more than 1 DLL from .libs:
c:\Users\sraja\Miniconda3\lib\site-packages\numpy\.libs\libopenblas.EL2C6PLE4ZYW3ECEVIV3OXXGRN2NRFM2.gfortran-win_amd64.dll
c:\Users\sraja\Miniconda3\lib\site-packages\numpy\.libs\libopenblas.XWYDX2IKJW2NMTWSFYNGFUWKQU3LYTCZ.gfortran-win_amd64.dll
  warnings.warn("loaded more than 1 DLL from .libs:"


In [2]:
df = df2009.append(df2010)
df

Invoice StockCode                          Description  Quantity  \
0       489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1       489434    79323P                   PINK CHERRY LIGHTS        12   
2       489434    79323W                  WHITE CHERRY LIGHTS        12   
3       489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4       489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...        ...       ...                                  ...       ...   
541905  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
541909  581587      POST                              POSTAGE         1   

             InvoiceDate  Price  Customer ID         Country  
0       01-12-2009 07:45   6.95      13085.0  United Kingdom  
1       01-12-2009 07:45   6.75      13085.0  United Kingdom  
2       01-12-2009 07:45   6.75      13085.0  United Kingdom  
3       01-12-2009 07:45   2.10      13085.0  United Kingdom  
4       01-12-2009 07:45   1.25      13085.0  United Kingdom  
...                  ...    ...          ...             ...  
541905  09-12-2011 12:50   2.10      12680.0          France  
541906  09-12-2011 12:50   4.15      12680.0          France  
541907  09-12-2011 12:50   4.15      12680.0          France  
541908  09-12-2011 12:50   4.95      12680.0          France  
541909  09-12-2011 12:50  18.00      12680.0          France  

[1067371 rows x 8 columns]

## Duplication handling


in the original incoming XLSX file , there are 2 sheets , one for 2009-10 and one for 2010-11
for these periods there is overlap in reocrds that are exactly same 

this absolute duplicates need to be removed first ( where all columns match )

---

at a later stage , after the rest of Data Cleaning , 

we will revist duplicates , and handle the entries which violate the level of the Data

( ie. Multiple invoice-item level entries )

for now just droppping instances that have absolute duplication across all columns  

In [4]:
df.drop_duplicates(subset=None , keep='first', inplace=True )

----

In [41]:
DataFrameSummary(df).summary()

Invoice_Raw   Invoice_No Invoice_Status Product_Code  \
count                 NaN          NaN            NaN          NaN   
mean                  NaN          NaN            NaN          NaN   
std                   NaN          NaN            NaN          NaN   
min                   NaN          NaN            NaN          NaN   
25%                   NaN          NaN            NaN          NaN   
50%                   NaN          NaN            NaN          NaN   
75%                   NaN          NaN            NaN          NaN   
max                   NaN          NaN            NaN          NaN   
counts            1033031      1033031        1033031      1033031   
uniques             53623        53623              3         5305   
missing                 0            0              0            0   
missing_perc           0%           0%             0%           0%   
types         categorical  categorical    categorical  categorical   

                  Cust_ID Invoice_Date Product_Qty Product_Price  \
count                 NaN          NaN   1033031.0     1033031.0   
mean                  NaN          NaN   10.076923      4.767605   
std                   NaN          NaN  175.198051     92.837349   
min                   NaN          NaN    -80995.0           0.0   
25%                   NaN          NaN         1.0          1.25   
50%                   NaN          NaN         3.0           2.1   
75%                   NaN          NaN        10.0          4.15   
max                   NaN          NaN     80995.0       38970.0   
counts            1033031      1033031     1033031       1033031   
uniques              5943        47631        1057          2803   
missing                 0            0           0             0   
missing_perc           0%           0%          0%            0%   
types         categorical         date     numeric       numeric   

             Misc_Product_Code_Flag Cust_Country Product_Desc     Sale_Value  
count                           NaN          NaN          NaN      1033031.0  
mean                            NaN          NaN          NaN      18.406233  
std                             NaN          NaN          NaN     284.718683  
min                             NaN          NaN          NaN -168469.593821  
25%                             NaN          NaN          NaN           3.75  
50%                             NaN          NaN          NaN           9.92  
75%                             NaN          NaN          NaN           17.7  
max                             NaN          NaN          NaN  168469.593821  
counts                      1033031      1033031      1033031        1033031  
uniques                           2           43         5699          10340  
missing                           0            0            0              0  
missing_perc                     0%           0%           0%             0%  
types                          bool  categorical  categorical        numeric

## Column Renaming
### for better readability and ease-of-use

In [5]:
df.columns.values.tolist()

['Invoice',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'Price',
 'Customer ID',
 'Country']

In [6]:
rename_dict = {
    'Invoice' : 'Invoice_Raw' ,
    'StockCode' : 'Product_Code' ,
    'Description' : 'Product_Desc' ,
    'Quantity' : 'Product_Qty' ,
    'InvoiceDate' : 'Invoice_Date' ,
    'Price' : 'Product_Price' ,
    'Customer ID' : 'Cust_ID' ,
    'Country' : 'Cust_Country' 
    }

df = df.rename( columns = rename_dict )
df

Invoice_Raw Product_Code                         Product_Desc  \
0           489434        85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS   
1           489434       79323P                   PINK CHERRY LIGHTS   
2           489434       79323W                  WHITE CHERRY LIGHTS   
3           489434        22041         RECORD FRAME 7" SINGLE SIZE    
4           489434        21232       STRAWBERRY CERAMIC TRINKET BOX   
...            ...          ...                                  ...   
541905      581587        22899         CHILDREN'S APRON DOLLY GIRL    
541906      581587        23254        CHILDRENS CUTLERY DOLLY GIRL    
541907      581587        23255      CHILDRENS CUTLERY CIRCUS PARADE   
541908      581587        22138        BAKING SET 9 PIECE RETROSPOT    
541909      581587         POST                              POSTAGE   

        Product_Qty      Invoice_Date  Product_Price  Cust_ID    Cust_Country  
0                12  01-12-2009 07:45           6.95  13085.0  United Kingdom  
1                12  01-12-2009 07:45           6.75  13085.0  United Kingdom  
2                12  01-12-2009 07:45           6.75  13085.0  United Kingdom  
3                48  01-12-2009 07:45           2.10  13085.0  United Kingdom  
4                24  01-12-2009 07:45           1.25  13085.0  United Kingdom  
...             ...               ...            ...      ...             ...  
541905            6  09-12-2011 12:50           2.10  12680.0          France  
541906            4  09-12-2011 12:50           4.15  12680.0          France  
541907            4  09-12-2011 12:50           4.15  12680.0          France  
541908            3  09-12-2011 12:50           4.95  12680.0          France  
541909            1  09-12-2011 12:50          18.00  12680.0          France  

[1033036 rows x 8 columns]

## Invoice_Code_Raw
### break down composite column in to simple ( uniform data type columns )


In [7]:
df['Invoice_No']    = df['Invoice_Raw'].str.extract('([0-9]\w{0,})', expand=False)
df['Invoice_Status']= df['Invoice_Raw'].str.extract('([A-Z])', expand=False).fillna('F')

In [ ]:
# df['Product_Code'].value_counts ()
# len(df['Product_Code'])

# df['Product_Code'][ ~df['Product_Code'].str.isdigit() ].value_counts ()
# len(df['Product_Code'][ ~df['Product_Code'].str.isdigit() ])

---
## Product Code ( StockCode ) deepdive  

1715 ( 32% ) Keycodes of 5305 Total keycodes , which have a non Numeric Keycode

134986 ( 12% ) instances ( Invoices ) of total 1 Million Instances , which have a non Numeric Keycode

at this time it might be better to not treat/process Product Keycode in any way without further information

the following are Prodcut Keycodes , which stand out from teh rest , 
and might be used for some other process ( non standard invoice entries for adustment , misc charges etc)


- BANK CHARGES
- POST
- DOT
- BANK CHARGES
- AMAZONFEE
- ADJUST
- D ( Discount )


It might be worth considering to create flag column for this special Keycodes , 
if a accurate list were available ,


For Any Product Dimension Analysis , these Keycodes should be Avoided



In [8]:
df['Misc_Product_Code_Flag'] = df['Product_Code'].isin([
        'BANK CHARGES',
        'POST',
        'DOT',
        'BANK CHARGES',
        'AMAZONFEE',
        'ADJUST',
        'D','M',
        'Adjust bad debt',

        # the below list comes from instaces of negative QTY and Negative Price , captured in teh code blocks below 
        'check', 'damages', '?', 'damaged', 'missing', 'sold as set on dotcom','Damaged', 'thrown away', 'Unsaleable, destroyed.', 'smashed', 'dotcom','??', 'damages?', 'given away', 'crushed', 'MIA', 'ebay', 'counted','wet damaged', 'Damages'
    ])

---
## Product Qty Negatives deepdive  

In [9]:
df[df['Product_Qty'] <= 0]

Invoice_Raw Product_Code                      Product_Desc  \
178        C489449        22087          PAPER BUNTING WHITE LACE   
179        C489449       85206A      CREAM FELT EASTER EGG BASKET   
180        C489449        21895     POTTING SHED SOW 'N' GROW SET   
181        C489449        21896                POTTING SHED TWINE   
182        C489449        22083        PAPER CHAIN KIT RETRO SPOT   
...            ...          ...                               ...   
540449     C581490        23144   ZINC T-LIGHT HOLDER STARS SMALL   
541541     C581499            M                            Manual   
541715     C581568        21258        VICTORIAN SEWING BOX LARGE   
541716     C581569        84978  HANGING HEART JAR T-LIGHT HOLDER   
541717     C581569        20979     36 PENCILS TUBE RED RETROSPOT   

        Product_Qty      Invoice_Date  Product_Price  Cust_ID    Cust_Country  \
178             -12  01-12-2009 10:33           2.95  16321.0       Australia   
179              -6  01-12-2009 10:33           1.65  16321.0       Australia   
180              -4  01-12-2009 10:33           4.25  16321.0       Australia   
181              -6  01-12-2009 10:33           2.10  16321.0       Australia   
182             -12  01-12-2009 10:33           2.95  16321.0       Australia   
...             ...               ...            ...      ...             ...   
540449          -11  09-12-2011 09:57           0.83  14397.0  United Kingdom   
541541           -1  09-12-2011 10:28         224.69  15498.0  United Kingdom   
541715           -5  09-12-2011 11:57          10.95  15311.0  United Kingdom   
541716           -1  09-12-2011 11:58           1.25  17315.0  United Kingdom   
541717           -5  09-12-2011 11:58           1.25  17315.0  United Kingdom   

       Invoice_No Invoice_Status  Misc_Product_Code_Flag  
178        489449              C                   False  
179        489449              C                   False  
180        489449              C                   False  
181        489449              C                   False  
182        489449              C                   False  
...           ...            ...                     ...  
540449     581490              C                   False  
541541     581499              C                    True  
541715     581568              C                   False  
541716     581569              C                   False  
541717     581569              C                   False  

[22496 rows x 11 columns]

In [10]:
df[ (df['Product_Qty'] <= 0 ) & (df['Invoice_Status'] != 'C' ) ]

Invoice_Raw Product_Code  Product_Desc  Product_Qty      Invoice_Date  \
263         489464        21733  85123a mixed          -96  01-12-2009 10:52   
283         489463        71477         short         -240  01-12-2009 10:52   
284         489467       85123A   21733 mixed         -192  01-12-2009 10:53   
470         489521        21646           NaN          -50  01-12-2009 11:44   
3114        489655        20683           NaN          -44  01-12-2009 17:26   
...            ...          ...           ...          ...               ...   
535333      581210        23395         check          -26  07-12-2011 18:36   
535335      581212        22578          lost        -1050  07-12-2011 18:38   
535336      581213        22576         check          -30  07-12-2011 18:38   
536910      581226        23090       missing         -338  08-12-2011 09:56   
538925      581422        23169       smashed         -235  08-12-2011 15:24   

        Product_Price  Cust_ID    Cust_Country Invoice_No Invoice_Status  \
263               0.0      NaN  United Kingdom     489464              F   
283               0.0      NaN  United Kingdom     489463              F   
284               0.0      NaN  United Kingdom     489467              F   
470               0.0      NaN  United Kingdom     489521              F   
3114              0.0      NaN  United Kingdom     489655              F   
...               ...      ...             ...        ...            ...   
535333            0.0      NaN  United Kingdom     581210              F   
535335            0.0      NaN  United Kingdom     581212              F   
535336            0.0      NaN  United Kingdom     581213              F   
536910            0.0      NaN  United Kingdom     581226              F   
538925            0.0      NaN  United Kingdom     581422              F   

        Misc_Product_Code_Flag  
263                      False  
283                      False  
284                      False  
470                      False  
3114                     False  
...                        ...  
535333                   False  
535335                   False  
535336                   False  
536910                   False  
538925                   False  

[3393 rows x 11 columns]

In [11]:
df[ (df['Product_Qty'] <= 0 ) & (df['Invoice_Status'] != 'C' )  & (~ df['Cust_ID'].isna() ) ]

Empty DataFrame
Columns: [Invoice_Raw, Product_Code, Product_Desc, Product_Qty, Invoice_Date, Product_Price, Cust_ID, Cust_Country, Invoice_No, Invoice_Status, Misc_Product_Code_Flag]
Index: []

so Price is negative only for instaces that are Canceled orders AND  Cust_ID is NA


---
## Product Price Negatives deepdive  

In [12]:
df[df['Product_Price'] <0]

Invoice_Raw Product_Code     Product_Desc  Product_Qty  \
179403     A506401            B  Adjust bad debt            1   
276274     A516228            B  Adjust bad debt            1   
403472     A528059            B  Adjust bad debt            1   
299983     A563186            B  Adjust bad debt            1   
299984     A563187            B  Adjust bad debt            1   

            Invoice_Date  Product_Price  Cust_ID    Cust_Country Invoice_No  \
179403  29-04-2010 13:36      -53594.36      NaN  United Kingdom     506401   
276274  19-07-2010 11:24      -44031.79      NaN  United Kingdom     516228   
403472  20-10-2010 12:04      -38925.87      NaN  United Kingdom     528059   
299983  12-08-2011 14:51      -11062.06      NaN  United Kingdom     563186   
299984  12-08-2011 14:52      -11062.06      NaN  United Kingdom     563187   

       Invoice_Status  Misc_Product_Code_Flag  
179403              A                   False  
276274              A                   False  
403472              A                   False  
299983              A                   False  
299984              A                   False

These are very heavy outliers and can screw up the entire analysis if mistakenly included , 

Since we are not conducting any finanical statement ( balance sheet ) type of ananlysis , it is safer to get rid of these completely by deleting this instances



In [13]:
df = df[df['Product_Price'] >= 0]

In [14]:
df[df['Product_Price'] == 0]

Invoice_Raw Product_Code                  Product_Desc  Product_Qty  \
263         489464        21733                  85123a mixed          -96   
283         489463        71477                         short         -240   
284         489467       85123A                   21733 mixed         -192   
470         489521        21646                           NaN          -50   
3114        489655        20683                           NaN          -44   
...            ...          ...                           ...          ...   
536981      581234        72817                           NaN           27   
538504      581406       46000M  POLYESTER FILLER PAD 45x45cm          240   
538505      581406       46000S  POLYESTER FILLER PAD 40x40cm          300   
538554      581408        85175                           NaN           20   
538925      581422        23169                       smashed         -235   

            Invoice_Date  Product_Price  Cust_ID    Cust_Country Invoice_No  \
263     01-12-2009 10:52            0.0      NaN  United Kingdom     489464   
283     01-12-2009 10:52            0.0      NaN  United Kingdom     489463   
284     01-12-2009 10:53            0.0      NaN  United Kingdom     489467   
470     01-12-2009 11:44            0.0      NaN  United Kingdom     489521   
3114    01-12-2009 17:26            0.0      NaN  United Kingdom     489655   
...                  ...            ...      ...             ...        ...   
536981  08-12-2011 10:33            0.0      NaN  United Kingdom     581234   
538504  08-12-2011 13:58            0.0      NaN  United Kingdom     581406   
538505  08-12-2011 13:58            0.0      NaN  United Kingdom     581406   
538554  08-12-2011 14:06            0.0      NaN  United Kingdom     581408   
538925  08-12-2011 15:24            0.0      NaN  United Kingdom     581422   

       Invoice_Status  Misc_Product_Code_Flag  
263                 F                   False  
283                 F                   False  
284                 F                   False  
470                 F                   False  
3114                F                   False  
...               ...                     ...  
536981              F                   False  
538504              F                   False  
538505              F                   False  
538554              F                   False  
538925              F                   False  

[6014 rows x 11 columns]

most of the Product_QTYs seem to be negative

In [15]:
df[ (df['Product_Qty'] == 0 ) & (df['Product_Price'] == 0) ]

Empty DataFrame
Columns: [Invoice_Raw, Product_Code, Product_Desc, Product_Qty, Invoice_Date, Product_Price, Cust_ID, Cust_Country, Invoice_No, Invoice_Status, Misc_Product_Code_Flag]
Index: []

In [16]:
df[ (df['Product_Qty'] > 0 ) & (df['Product_Price'] == 0) ]

Invoice_Raw Product_Code                  Product_Desc  Product_Qty  \
3161        489659        21350                           NaN          230   
3731        489781        84292                           NaN           17   
4674        489825        22076            6 RIBBONS EMPIRE             12   
5904        489861          DOT                DOTCOM POSTAGE            1   
6378        489882       35751C                           NaN           12   
...            ...          ...                           ...          ...   
535334      581211        22142                         check           14   
536981      581234        72817                           NaN           27   
538504      581406       46000M  POLYESTER FILLER PAD 45x45cm          240   
538505      581406       46000S  POLYESTER FILLER PAD 40x40cm          300   
538554      581408        85175                           NaN           20   

            Invoice_Date  Product_Price  Cust_ID    Cust_Country Invoice_No  \
3161    01-12-2009 17:39            0.0      NaN  United Kingdom     489659   
3731    02-12-2009 11:45            0.0      NaN  United Kingdom     489781   
4674    02-12-2009 13:34            0.0  16126.0  United Kingdom     489825   
5904    02-12-2009 14:50            0.0      NaN  United Kingdom     489861   
6378    02-12-2009 16:22            0.0      NaN  United Kingdom     489882   
...                  ...            ...      ...             ...        ...   
535334  07-12-2011 18:36            0.0      NaN  United Kingdom     581211   
536981  08-12-2011 10:33            0.0      NaN  United Kingdom     581234   
538504  08-12-2011 13:58            0.0      NaN  United Kingdom     581406   
538505  08-12-2011 13:58            0.0      NaN  United Kingdom     581406   
538554  08-12-2011 14:06            0.0      NaN  United Kingdom     581408   

       Invoice_Status  Misc_Product_Code_Flag  
3161                F                   False  
3731                F                   False  
4674                F                   False  
5904                F                    True  
6378                F                   False  
...               ...                     ...  
535334              F                   False  
536981              F                   False  
538504              F                   False  
538505              F                   False  
538554              F                   False  

[2621 rows x 11 columns]

In [17]:
df[ (df['Product_Qty'] < 0 ) & (df['Product_Price'] == 0) ]

Invoice_Raw Product_Code  Product_Desc  Product_Qty      Invoice_Date  \
263         489464        21733  85123a mixed          -96  01-12-2009 10:52   
283         489463        71477         short         -240  01-12-2009 10:52   
284         489467       85123A   21733 mixed         -192  01-12-2009 10:53   
470         489521        21646           NaN          -50  01-12-2009 11:44   
3114        489655        20683           NaN          -44  01-12-2009 17:26   
...            ...          ...           ...          ...               ...   
535333      581210        23395         check          -26  07-12-2011 18:36   
535335      581212        22578          lost        -1050  07-12-2011 18:38   
535336      581213        22576         check          -30  07-12-2011 18:38   
536910      581226        23090       missing         -338  08-12-2011 09:56   
538925      581422        23169       smashed         -235  08-12-2011 15:24   

        Product_Price  Cust_ID    Cust_Country Invoice_No Invoice_Status  \
263               0.0      NaN  United Kingdom     489464              F   
283               0.0      NaN  United Kingdom     489463              F   
284               0.0      NaN  United Kingdom     489467              F   
470               0.0      NaN  United Kingdom     489521              F   
3114              0.0      NaN  United Kingdom     489655              F   
...               ...      ...             ...        ...            ...   
535333            0.0      NaN  United Kingdom     581210              F   
535335            0.0      NaN  United Kingdom     581212              F   
535336            0.0      NaN  United Kingdom     581213              F   
536910            0.0      NaN  United Kingdom     581226              F   
538925            0.0      NaN  United Kingdom     581422              F   

        Misc_Product_Code_Flag  
263                      False  
283                      False  
284                      False  
470                      False  
3114                     False  
...                        ...  
535333                   False  
535335                   False  
535336                   False  
536910                   False  
538925                   False  

[3393 rows x 11 columns]

best would be to mark these as True for the Misc_Product_Code_Flag column

these product_desc have been added to the list in the above code blocks

-----

# dealing with NULL values in Cust_ID

ignoring NULLs in Product Description as it is a non Critical column 

In [18]:
df.isna().sum()

Invoice_Raw                    0
Product_Code                   0
Product_Desc                4275
Product_Qty                    0
Invoice_Date                   0
Product_Price                  0
Cust_ID                   235146
Cust_Country                   0
Invoice_No                     0
Invoice_Status                 0
Misc_Product_Code_Flag         0
dtype: int64

In [19]:
df.dtypes

Invoice_Raw                object
Product_Code               object
Product_Desc               object
Product_Qty                 int64
Invoice_Date               object
Product_Price             float64
Cust_ID                   float64
Cust_Country               object
Invoice_No                 object
Invoice_Status             object
Misc_Product_Code_Flag       bool
dtype: object

In [20]:
df [ df['Cust_ID'].isna() ] #.to_clipboard()

Invoice_Raw Product_Code                    Product_Desc  Product_Qty  \
263         489464        21733                    85123a mixed          -96   
283         489463        71477                           short         -240   
284         489467       85123A                     21733 mixed         -192   
470         489521        21646                             NaN          -50   
577         489525       85226C       BLUE PULL BACK RACING CAR            1   
...            ...          ...                             ...          ...   
541536      581498       85099B         JUMBO BAG RED RETROSPOT            5   
541537      581498       85099C  JUMBO  BAG BAROQUE BLACK WHITE            4   
541538      581498        85150   LADIES & GENTLEMEN METAL SIGN            1   
541539      581498        85174               S/4 CACTI CANDLES            1   
541540      581498          DOT                  DOTCOM POSTAGE            1   

            Invoice_Date  Product_Price  Cust_ID    Cust_Country Invoice_No  \
263     01-12-2009 10:52           0.00      NaN  United Kingdom     489464   
283     01-12-2009 10:52           0.00      NaN  United Kingdom     489463   
284     01-12-2009 10:53           0.00      NaN  United Kingdom     489467   
470     01-12-2009 11:44           0.00      NaN  United Kingdom     489521   
577     01-12-2009 11:49           0.55      NaN  United Kingdom     489525   
...                  ...            ...      ...             ...        ...   
541536  09-12-2011 10:26           4.13      NaN  United Kingdom     581498   
541537  09-12-2011 10:26           4.13      NaN  United Kingdom     581498   
541538  09-12-2011 10:26           4.96      NaN  United Kingdom     581498   
541539  09-12-2011 10:26          10.79      NaN  United Kingdom     581498   
541540  09-12-2011 10:26        1714.17      NaN  United Kingdom     581498   

       Invoice_Status  Misc_Product_Code_Flag  
263                 F                   False  
283                 F                   False  
284                 F                   False  
470                 F                   False  
577                 F                   False  
...               ...                     ...  
541536              F                   False  
541537              F                   False  
541538              F                   False  
541539              F                   False  
541540              F                    True  

[235146 rows x 11 columns]


can impute customer ID from other matching Invoice Numbers 

as there is a one way distinct mapping ( one to many ) for customers to invoices 

Ie one customer can have many invoices , but one invoice is associated with only a single customer

so can use Repeating Invoice Numbers to get Customer ID

need to create a Customer-Invoice Map , and then use that to impute teh missing NaNs

In [21]:
df['Cust_ID'].isna().sum()

235146

In [26]:
invoice_customer_mapping = df[['Invoice_No','Cust_ID']] \
                            .dropna( how = 'any' ) \
                            .groupby( by = 'Invoice_No').agg(lambda x : x.value_counts().index[0] ) \
                            .to_dict( 'dict' )['Cust_ID']

df['Cust_ID'] = df['Cust_ID'].fillna( df['Invoice_No'].map(invoice_customer_mapping) )

C:\Users\sraja\AppData\Local\Temp/ipykernel_28744/379585267.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Cust_ID'] = df['Cust_ID'].fillna( df['Invoice_No'].map(invoice_customer_mapping) )


In [27]:
invoice_customer_mapping = df[['Invoice_No','Cust_ID']] \
                            .dropna( how = 'any' ) \
                            .groupby( by = 'Invoice_No').agg(pd.Series.mode) \
                            .to_dict( 'dict' )['Cust_ID']

df['Cust_ID'] = df['Cust_ID'].fillna( df['Invoice_No'].map(invoice_customer_mapping) )

C:\Users\sraja\AppData\Local\Temp/ipykernel_28744/3530146141.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['Cust_ID'] = df['Cust_ID'].fillna( df['Invoice_No'].map(invoice_customer_mapping) )


In [28]:
df['Cust_ID'].isna().sum()

235146

In [29]:
df['Cust_ID'].nunique()

5942

This method is proven to not be useful ,

because the Null Cust_ID values did get imputed by the rest of the Invoice-Customer Data

:(

----

## Re-order the columns 
### for ease of use

In [30]:
df.columns.values.tolist()

['Invoice_Raw',
 'Product_Code',
 'Product_Desc',
 'Product_Qty',
 'Invoice_Date',
 'Product_Price',
 'Cust_ID',
 'Cust_Country',
 'Invoice_No',
 'Invoice_Status',
 'Misc_Product_Code_Flag']

In [31]:
df = df [[
    
    # Primary Identifier - Level Column
    'Invoice_Raw',
    'Invoice_No',
    'Invoice_Status',
    'Product_Code',
    
    # Information Columns
    'Cust_ID',
    'Invoice_Date',    
    'Product_Qty',
    'Product_Price',
    'Misc_Product_Code_Flag',
    'Cust_Country' ,
    'Product_Desc',

]]
df

Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
0           489434     489434              F        85048  13085.0   
1           489434     489434              F       79323P  13085.0   
2           489434     489434              F       79323W  13085.0   
3           489434     489434              F        22041  13085.0   
4           489434     489434              F        21232  13085.0   
...            ...        ...            ...          ...      ...   
541905      581587     581587              F        22899  12680.0   
541906      581587     581587              F        23254  12680.0   
541907      581587     581587              F        23255  12680.0   
541908      581587     581587              F        22138  12680.0   
541909      581587     581587              F         POST  12680.0   

            Invoice_Date  Product_Qty  Product_Price  Misc_Product_Code_Flag  \
0       01-12-2009 07:45           12           6.95                   False   
1       01-12-2009 07:45           12           6.75                   False   
2       01-12-2009 07:45           12           6.75                   False   
3       01-12-2009 07:45           48           2.10                   False   
4       01-12-2009 07:45           24           1.25                   False   
...                  ...          ...            ...                     ...   
541905  09-12-2011 12:50            6           2.10                   False   
541906  09-12-2011 12:50            4           4.15                   False   
541907  09-12-2011 12:50            4           4.15                   False   
541908  09-12-2011 12:50            3           4.95                   False   
541909  09-12-2011 12:50            1          18.00                    True   

          Cust_Country                         Product_Desc  
0       United Kingdom  15CM CHRISTMAS GLASS BALL 20 LIGHTS  
1       United Kingdom                   PINK CHERRY LIGHTS  
2       United Kingdom                  WHITE CHERRY LIGHTS  
3       United Kingdom         RECORD FRAME 7" SINGLE SIZE   
4       United Kingdom       STRAWBERRY CERAMIC TRINKET BOX  
...                ...                                  ...  
541905          France         CHILDREN'S APRON DOLLY GIRL   
541906          France        CHILDRENS CUTLERY DOLLY GIRL   
541907          France      CHILDRENS CUTLERY CIRCUS PARADE  
541908          France        BAKING SET 9 PIECE RETROSPOT   
541909          France                              POSTAGE  

[1033031 rows x 11 columns]

## Type Setting of the Columns

In [32]:
df = df.astype({
    'Invoice_No'        : 'str' ,
    'Invoice_Status'    : 'str' ,
    'Invoice_Raw'       : 'str' ,
    'Invoice_Date'      : 'datetime64[ns]' ,
    'Product_Code'      : 'str' ,
    'Product_Desc'      : 'str' ,
    'Product_Qty'       : 'int32' ,
    'Product_Price'     : 'float32' ,
    'Product_Desc'      : 'str',
    'Cust_ID'           : 'str' ,
    'Cust_Country'      : 'str',
    'Misc_Product_Code_Flag' : 'str',
    })
df.dtypes

Invoice_Raw                       object
Invoice_No                        object
Invoice_Status                    object
Product_Code                      object
Cust_ID                           object
Invoice_Date              datetime64[ns]
Product_Qty                        int32
Product_Price                    float32
Misc_Product_Code_Flag            object
Cust_Country                      object
Product_Desc                      object
dtype: object

# Data level Check
### we expect that the Data is at Invoice-Item level

## Level Violations
entries that duplicate on the level of the data

In [33]:
df

Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
0           489434     489434              F        85048  13085.0   
1           489434     489434              F       79323P  13085.0   
2           489434     489434              F       79323W  13085.0   
3           489434     489434              F        22041  13085.0   
4           489434     489434              F        21232  13085.0   
...            ...        ...            ...          ...      ...   
541905      581587     581587              F        22899  12680.0   
541906      581587     581587              F        23254  12680.0   
541907      581587     581587              F        23255  12680.0   
541908      581587     581587              F        22138  12680.0   
541909      581587     581587              F         POST  12680.0   

              Invoice_Date  Product_Qty  Product_Price Misc_Product_Code_Flag  \
0      2009-01-12 07:45:00           12           6.95                  False   
1      2009-01-12 07:45:00           12           6.75                  False   
2      2009-01-12 07:45:00           12           6.75                  False   
3      2009-01-12 07:45:00           48           2.10                  False   
4      2009-01-12 07:45:00           24           1.25                  False   
...                    ...          ...            ...                    ...   
541905 2011-09-12 12:50:00            6           2.10                  False   
541906 2011-09-12 12:50:00            4           4.15                  False   
541907 2011-09-12 12:50:00            4           4.15                  False   
541908 2011-09-12 12:50:00            3           4.95                  False   
541909 2011-09-12 12:50:00            1          18.00                   True   

          Cust_Country                         Product_Desc  
0       United Kingdom  15CM CHRISTMAS GLASS BALL 20 LIGHTS  
1       United Kingdom                   PINK CHERRY LIGHTS  
2       United Kingdom                  WHITE CHERRY LIGHTS  
3       United Kingdom         RECORD FRAME 7" SINGLE SIZE   
4       United Kingdom       STRAWBERRY CERAMIC TRINKET BOX  
...                ...                                  ...  
541905          France         CHILDREN'S APRON DOLLY GIRL   
541906          France        CHILDRENS CUTLERY DOLLY GIRL   
541907          France      CHILDRENS CUTLERY CIRCUS PARADE  
541908          France        BAKING SET 9 PIECE RETROSPOT   
541909          France                              POSTAGE  

[1033031 rows x 11 columns]

In [34]:
level_test_list = [
    # 'Invoice_Raw',
    'Invoice_No',
    # 'Invoice_Status',
    'Product_Code',
    # 'Cust_ID',
    # 'Invoice_Date',
    # 'Product_Qty',
    # 'Product_Price',
    # 'Misc_Product_Code_Flag',
    # 'Cust_Country',
    # 'Product_Desc'
    ]

temp_level = df[ df.duplicated( subset = level_test_list , keep = False ) ].sort_values( by = level_test_list )

print( 'Key Duplications ( Level Violations ) % : ' , len(temp_level) / len(df) *100 )

temp_level

Key Duplications ( Level Violations ) % :  2.2045805014563937


Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
289         489488     489488              F        22125  17238.0   
299         489488     489488              F        22125  17238.0   
373         489517     489517              F        21790  16329.0   
393         489517     489517              F        21790  16329.0   
372         489517     489517              F        21791  16329.0   
...            ...        ...            ...          ...      ...   
541680      581538     581538              F        23349  14446.0   
541674      581538     581538              F       35004B  14446.0   
541698      581538     581538              F       35004B  14446.0   
541642      581538     581538              F        84380  14446.0   
541665      581538     581538              F        84380  14446.0   

              Invoice_Date  Product_Qty  Product_Price Misc_Product_Code_Flag  \
289    2009-01-12 10:59:00            2           5.95                  False   
299    2009-01-12 10:59:00            1           5.95                  False   
373    2009-01-12 11:34:00            4           0.85                  False   
393    2009-01-12 11:34:00            1           0.85                  False   
372    2009-01-12 11:34:00            6           1.25                  False   
...                    ...          ...            ...                    ...   
541680 2011-09-12 11:34:00            1           1.25                  False   
541674 2011-09-12 11:34:00            2           5.45                  False   
541698 2011-09-12 11:34:00            1           5.45                  False   
541642 2011-09-12 11:34:00            1           1.45                  False   
541665 2011-09-12 11:34:00            4           1.45                  False   

          Cust_Country                        Product_Desc  
289     United Kingdom        UNION JACK HOT WATER BOTTLE   
299     United Kingdom        UNION JACK HOT WATER BOTTLE   
373     United Kingdom                  VINTAGE SNAP CARDS  
393     United Kingdom                  VINTAGE SNAP CARDS  
372     United Kingdom  VINTAGE HEADS AND TAILS CARD GAME   
...                ...                                 ...  
541680  United Kingdom         ROLL WRAP VINTAGE CHRISTMAS  
541674  United Kingdom         SET OF 3 BLACK FLYING DUCKS  
541698  United Kingdom         SET OF 3 BLACK FLYING DUCKS  
541642  United Kingdom   SET OF 3 BUTTERFLY COOKIE CUTTERS  
541665  United Kingdom   SET OF 3 BUTTERFLY COOKIE CUTTERS  

[22774 rows x 11 columns]

there seems to be multiple entries ( 22K ) for the same invoice with different prodcut_qty s

this can be handled with Aggreagtion , by taking Sum



In [35]:
level_test_list = [
    # 'Invoice_Raw',
    'Invoice_No',
    'Invoice_Status',
    'Product_Code',
    # 'Cust_ID',
    # 'Invoice_Date',
    # 'Product_Qty',
    # 'Product_Price',
    # 'Misc_Product_Code_Flag',
    # 'Cust_Country',
    # 'Product_Desc'
    ]

temp_level = df[ df.duplicated( subset = level_test_list , keep = False ) ].sort_values( by = level_test_list )

print( 'Key Duplications ( Level Violations ) % : ' , len(temp_level) / len(df) *100 )

temp_level

Key Duplications ( Level Violations ) % :  2.2045805014563937


Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
289         489488     489488              F        22125  17238.0   
299         489488     489488              F        22125  17238.0   
373         489517     489517              F        21790  16329.0   
393         489517     489517              F        21790  16329.0   
372         489517     489517              F        21791  16329.0   
...            ...        ...            ...          ...      ...   
541680      581538     581538              F        23349  14446.0   
541674      581538     581538              F       35004B  14446.0   
541698      581538     581538              F       35004B  14446.0   
541642      581538     581538              F        84380  14446.0   
541665      581538     581538              F        84380  14446.0   

              Invoice_Date  Product_Qty  Product_Price Misc_Product_Code_Flag  \
289    2009-01-12 10:59:00            2           5.95                  False   
299    2009-01-12 10:59:00            1           5.95                  False   
373    2009-01-12 11:34:00            4           0.85                  False   
393    2009-01-12 11:34:00            1           0.85                  False   
372    2009-01-12 11:34:00            6           1.25                  False   
...                    ...          ...            ...                    ...   
541680 2011-09-12 11:34:00            1           1.25                  False   
541674 2011-09-12 11:34:00            2           5.45                  False   
541698 2011-09-12 11:34:00            1           5.45                  False   
541642 2011-09-12 11:34:00            1           1.45                  False   
541665 2011-09-12 11:34:00            4           1.45                  False   

          Cust_Country                        Product_Desc  
289     United Kingdom        UNION JACK HOT WATER BOTTLE   
299     United Kingdom        UNION JACK HOT WATER BOTTLE   
373     United Kingdom                  VINTAGE SNAP CARDS  
393     United Kingdom                  VINTAGE SNAP CARDS  
372     United Kingdom  VINTAGE HEADS AND TAILS CARD GAME   
...                ...                                 ...  
541680  United Kingdom         ROLL WRAP VINTAGE CHRISTMAS  
541674  United Kingdom         SET OF 3 BLACK FLYING DUCKS  
541698  United Kingdom         SET OF 3 BLACK FLYING DUCKS  
541642  United Kingdom   SET OF 3 BUTTERFLY COOKIE CUTTERS  
541665  United Kingdom   SET OF 3 BUTTERFLY COOKIE CUTTERS  

[22774 rows x 11 columns]

there are no multiple entries for same order ID , 

for instance , with one of them being a Cancelled entry , while the other being the original one

In [36]:
level_test_list = [
    # 'Invoice_Raw',
    'Invoice_No',
    # 'Invoice_Status',
    'Product_Code',
    # 'Cust_ID',
    # 'Invoice_Date',
    'Product_Qty',
    # 'Product_Price',
    # 'Misc_Product_Code_Flag',
    # 'Cust_Country',
    # 'Product_Desc'
    ]

temp_level = df[ df.duplicated( subset = level_test_list , keep = False ) ].sort_values( by = level_test_list )

print( 'Key Duplications ( Level Violations ) % : ' , len(temp_level) / len(df) *100 )

temp_level

Key Duplications ( Level Violations ) % :  0.045787590111042165


Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
735        C489535     489535              C            D  15299.0   
736        C489535     489535              C            D  15299.0   
5064        489856     489856              F        22111      nan   
5080        489856     489856              F        22111      nan   
11310       490300     490300              F            M  12970.0   
...            ...        ...            ...          ...      ...   
529860      580754     580754              F        23168      nan   
537385      581256     581256              F        23168      nan   
537386      581256     581256              F        23168      nan   
541272      581497     581497              F        21481      nan   
541273      581497     581497              F        21481      nan   

              Invoice_Date  Product_Qty  Product_Price Misc_Product_Code_Flag  \
735    2009-01-12 12:11:00           -1           9.00                   True   
736    2009-01-12 12:11:00           -1          19.00                   True   
5064   2009-02-12 14:36:00            5           8.65                  False   
5080   2009-02-12 14:36:00            5          11.26                  False   
11310  2009-04-12 14:19:00            1           0.85                   True   
...                    ...          ...            ...                    ...   
529860 2011-06-12 10:05:00            1           6.54                  False   
537385 2011-08-12 11:21:00            1           3.29                  False   
537386 2011-08-12 11:21:00            1           6.54                  False   
541272 2011-09-12 10:23:00            1           8.29                  False   
541273 2011-09-12 10:23:00            1          10.29                  False   

          Cust_Country                  Product_Desc  
735     United Kingdom                      Discount  
736     United Kingdom                      Discount  
5064    United Kingdom  SCOTTIE DOG HOT WATER BOTTLE  
5080    United Kingdom  SCOTTIE DOG HOT WATER BOTTLE  
11310   United Kingdom                        Manual  
...                ...                           ...  
529860  United Kingdom  CLASSIC CAFE SUGAR DISPENSER  
537385  United Kingdom  CLASSIC CAFE SUGAR DISPENSER  
537386  United Kingdom  CLASSIC CAFE SUGAR DISPENSER  
541272  United Kingdom    FAWN BLUE HOT WATER BOTTLE  
541273  United Kingdom    FAWN BLUE HOT WATER BOTTLE  

[473 rows x 11 columns]

here Price seems to be the issue that is causing the duplicates , 

this can be handled with Aggreagtion , by taking Mean Price

lets see what happens If i add that into the lvel to check :

In [37]:
level_test_list = [
    # 'Invoice_Raw',
    'Invoice_No',
    # 'Invoice_Status',
    'Product_Code',
    # 'Cust_ID',
    # 'Invoice_Date',
    'Product_Qty',
    'Product_Price',
    # 'Misc_Product_Code_Flag',
    # 'Cust_Country',
    # 'Product_Desc'
    ]

temp_level = df[ df.duplicated( subset = level_test_list , keep = False ) ].sort_values( by = level_test_list )

print( 'Key Duplications ( Level Violations ) % : ' , len(temp_level) / len(df) *100 )

temp_level

Key Duplications ( Level Violations ) % :  0.0007744201287279859


Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
41637       492807     492807              F        21813  17211.0   
41657       492807     492807              F        21813  17211.0   
199124      554084     554084              F        23298  12909.0   
199144      554084     554084              F        23298  12909.0   
346033      567183     567183              F        22659  14769.0   
346281      567183     567183              F        22659  14769.0   
452368      575335     575335              F        23203  12931.0   
452371      575335     575335              F        23203  12931.0   

              Invoice_Date  Product_Qty  Product_Price Misc_Product_Code_Flag  \
41637  2009-12-20 12:28:00            6           4.95                  False   
41657  2009-12-20 12:29:00            6           4.95                  False   
199124 2011-05-22 11:52:00            3           4.95                  False   
199144 2011-05-22 11:52:00            3           4.95                  False   
346033 2011-09-18 15:32:00            2           1.95                  False   
346281 2011-09-18 15:33:00            2           1.95                  False   
452368 2011-09-11 13:56:00          300           1.79                  False   
452371 2011-09-11 13:56:00          300           1.79                  False   

          Cust_Country                  Product_Desc  
41637   United Kingdom  GARLAND WITH STARS AND BELLS  
41657   United Kingdom  GARLAND WITH STARS AND BELLS  
199124  United Kingdom             BUNTING , SPOTTY   
199144  United Kingdom                SPOTTY BUNTING  
346033  United Kingdom       LUNCH BOX I LOVE LONDON  
346281  United Kingdom       LUNCH BOX I LOVE LONDON  
452368  United Kingdom     JUMBO BAG DOILEY PATTERNS  
452371  United Kingdom      JUMBO BAG VINTAGE DOILY

the Final Descrepancy is in PRODUCT DESC

which we can ignore , as it doesn't have any material impact on the 1M large data set 
and there is not additional information captured as such anyway

----

# Final Level Fix ( Aggergation )

In [38]:
df['Sale_Value'] = df['Product_Qty'] * df['Product_Price'] 

In [39]:
df

Invoice_Raw Invoice_No Invoice_Status Product_Code  Cust_ID  \
0           489434     489434              F        85048  13085.0   
1           489434     489434              F       79323P  13085.0   
2           489434     489434              F       79323W  13085.0   
3           489434     489434              F        22041  13085.0   
4           489434     489434              F        21232  13085.0   
...            ...        ...            ...          ...      ...   
541905      581587     581587              F        22899  12680.0   
541906      581587     581587              F        23254  12680.0   
541907      581587     581587              F        23255  12680.0   
541908      581587     581587              F        22138  12680.0   
541909      581587     581587              F         POST  12680.0   

              Invoice_Date  Product_Qty  Product_Price Misc_Product_Code_Flag  \
0      2009-01-12 07:45:00           12           6.95                  False   
1      2009-01-12 07:45:00           12           6.75                  False   
2      2009-01-12 07:45:00           12           6.75                  False   
3      2009-01-12 07:45:00           48           2.10                  False   
4      2009-01-12 07:45:00           24           1.25                  False   
...                    ...          ...            ...                    ...   
541905 2011-09-12 12:50:00            6           2.10                  False   
541906 2011-09-12 12:50:00            4           4.15                  False   
541907 2011-09-12 12:50:00            4           4.15                  False   
541908 2011-09-12 12:50:00            3           4.95                  False   
541909 2011-09-12 12:50:00            1          18.00                   True   

          Cust_Country                         Product_Desc  Sale_Value  
0       United Kingdom  15CM CHRISTMAS GLASS BALL 20 LIGHTS   83.399998  
1       United Kingdom                   PINK CHERRY LIGHTS   81.000000  
2       United Kingdom                  WHITE CHERRY LIGHTS   81.000000  
3       United Kingdom         RECORD FRAME 7" SINGLE SIZE   100.799995  
4       United Kingdom       STRAWBERRY CERAMIC TRINKET BOX   30.000000  
...                ...                                  ...         ...  
541905          France         CHILDREN'S APRON DOLLY GIRL    12.599999  
541906          France        CHILDRENS CUTLERY DOLLY GIRL    16.600000  
541907          France      CHILDRENS CUTLERY CIRCUS PARADE   16.600000  
541908          France        BAKING SET 9 PIECE RETROSPOT    14.849999  
541909          France                              POSTAGE   18.000000  

[1033031 rows x 12 columns]

In [40]:
final_columns = [
'Invoice_No',
'Product_Code',
'Cust_ID',
'Invoice_Date',
'Invoice_Status',
'Product_Qty',
'Product_Price',
'Sale_Value',
'Misc_Product_Code_Flag',
'Cust_Country',
'Product_Desc',
]

# mode = lambda x : x.value_counts().index[0]
# mode = lambda x : x.mode(axis=0, numeric_only=False, dropna=False)

# mode did not work , because of pandas doesnt have a implementation , and the hacks neede dtoo much time to compute 
# used max instead , because anyways the columns have a single value , so mode and max would return the same values

final_df = df[final_columns].groupby(['Invoice_No','Product_Code']).agg(
    {
        'Cust_ID'                   : 'max' ,
        'Invoice_Date'              : 'min' ,
        'Invoice_Status'            : 'max' ,
        'Product_Qty'               : 'sum' ,
        'Product_Price'             : 'mean',
        'Sale_Value'                : 'sum' ,
        'Misc_Product_Code_Flag'    : 'max' ,
        'Cust_Country'              : 'max' ,
        'Product_Desc'              : 'max' ,
    }
)

final_df = final_df.reset_index()
final_df

Invoice_No Product_Code  Cust_ID        Invoice_Date Invoice_Status  \
0           489434        21232  13085.0 2009-01-12 07:45:00              F   
1           489434        21523  13085.0 2009-01-12 07:45:00              F   
2           489434        21871  13085.0 2009-01-12 07:45:00              F   
3           489434        22041  13085.0 2009-01-12 07:45:00              F   
4           489434        22064  13085.0 2009-01-12 07:45:00              F   
...            ...          ...      ...                 ...            ...   
1021414     581587        22899  12680.0 2011-09-12 12:50:00              F   
1021415     581587        23254  12680.0 2011-09-12 12:50:00              F   
1021416     581587        23255  12680.0 2011-09-12 12:50:00              F   
1021417     581587        23256  12680.0 2011-09-12 12:50:00              F   
1021418     581587         POST  12680.0 2011-09-12 12:50:00              F   

         Product_Qty  Product_Price  Sale_Value Misc_Product_Code_Flag  \
0                 24           1.25   30.000000                  False   
1                 10           5.95   59.499998                  False   
2                 24           1.25   30.000000                  False   
3                 48           2.10  100.799995                  False   
4                 24           1.65   39.599999                  False   
...              ...            ...         ...                    ...   
1021414            6           2.10   12.599999                  False   
1021415            4           4.15   16.600000                  False   
1021416            4           4.15   16.600000                  False   
1021417            4           4.15   16.600000                  False   
1021418            1          18.00   18.000000                   True   

           Cust_Country                        Product_Desc  
0        United Kingdom      STRAWBERRY CERAMIC TRINKET BOX  
1        United Kingdom  FANCY FONT HOME SWEET HOME DOORMAT  
2        United Kingdom                 SAVE THE PLANET MUG  
3        United Kingdom        RECORD FRAME 7" SINGLE SIZE   
4        United Kingdom          PINK DOUGHNUT TRINKET POT   
...                 ...                                 ...  
1021414          France        CHILDREN'S APRON DOLLY GIRL   
1021415          France       CHILDRENS CUTLERY DOLLY GIRL   
1021416          France     CHILDRENS CUTLERY CIRCUS PARADE  
1021417          France         CHILDRENS CUTLERY SPACEBOY   
1021418          France                             POSTAGE  

[1021419 rows x 11 columns]

In [41]:
final_df.to_csv('data_cleaned\\Data_1-Cleaned.csv' , index = False , header = True )

## Data Enhancement
### Auxilliary Features

----

For the countries column , 
there are 43 different countries 
of which  95% of the Transaction counts having only one Country UK

sicne this columnis heavily skewed over the dfset 

i have manually grouped the countries , based on the continents and valuecounts

In [42]:
df = final_df

In [43]:
# to export the df to excel
#df['Cust_Country'].value_counts().to_clipboard()
# Please refer exce file : Country_Region_MAP

country_df = pd.read_csv('Country_Region_MAP.csv')
country_df

df = df.merge( country_df , on='Cust_Country' , how='left' )

----

Next we also need to create some Auxillary columns for doing Time Dimension analysis


In [44]:
df['Invoice_Date']  = df['Invoice_Date'].astype('datetime64[ns]')
df["Year"]          = df['Invoice_Date'].dt.year
df["Quarter"]       = df['Invoice_Date'].dt.quarter
df["Month"]         = df['Invoice_Date'].dt.month
df["Week"]          = df['Invoice_Date'].dt.week
df["Weekday"]       = df['Invoice_Date'].dt.weekday
df["Day"]           = df['Invoice_Date'].dt.day
df["Date"]          = pd.to_datetime(df[['Year', 'Month', 'Day']])
df["Hour"]          = df['Invoice_Date'].dt.hour
df["Day_of_Year"]   = df['Invoice_Date'].dt.dayofyear
df["Year_Quarter"]  = df['Year']*10 + df['Quarter']
df["Year_Month"]    = df['Year']*100 + df['Month']
df["Year_Week"]     = df['Year']*100 + df['Week']
df["Year_Day"]      = df['Year']*1000 + df['Invoice_Date'].dt.dayofyear

C:\Users\sraja\AppData\Local\Temp/ipykernel_40480/4081696336.py:5: FutureWarning: Series.dt.weekofyear and Series.dt.week have been deprecated.  Please use Series.dt.isocalendar().week instead.
  df["Week"]          = df['Invoice_Date'].dt.week


In [ ]:
df.to_csv('data_cleaned\\Data_2-Enhanced.csv' , index = False , header = True )